gusto/diagnostics.py

```diff
@@ -1689,6 +1692,121 @@ def setup(self, domain, state_fields):
         super().setup(domain, state_fields, vorticity_type="relative")
 
 
+class Vorticity(DiagnosticField):
+    u""" Base diagnostic Field for three dimensional Vorticity """
+
+    def setup(self, domain, state_fields, vorticity_type=None):
+        """
+        Sets up the :class:`Function` for the diagnostic field.
+
+        Args:
+            domain (:class:`Domain`): the model's domain object.
+            state_fields (:class:`StateFields`): the model's field container.
+            vorticity_type (str, optional): denotes which type of vorticity to
+                be computed ('relative', 'absolute'). Defaults to
+                None.
+        """
+        # TODO Do we eventually want a potential voriticy?
+        vorticity_types = ['relative', 'absolute']
+        if vorticity_type not in vorticity_types:
+            if vorticity_type == 'potential':
+                raise ValueError('Potential vorticity has not yet been implemented')
+            else:
+                raise ValueError(f'vorticity type must be one of {vorticity_types}, not {vorticity_type}')
+        if domain.mesh.topological_dimension() == 3:
+            space = domain.spaces('HCurl')
+        else:
+            space = domain.spaces('H1')
+
+        u = state_fields('u')
+        if self.method != 'solve':
+            if vorticity_type == 'relative':
+                self.expression = curl(u)
+            elif vorticity_type == 'absolute':
+                Omega = as_vector((0, 0, self.parameters.Omega))
+                self.expression = curl(u) + 2*Omega
+
+        super().setup(domain, state_fields, space=space)
+
+        if self.method == 'solve':
+            if domain.mesh.topological_dimension() == 3:
+                vort = TrialFunction(space)
+                w = TestFunction(space)
+                n = FacetNormal(domain.mesh)
+                a = inner(vort, w) * dx
+                L = inner(u, curl(w)) * dx - jump(cross(w, u), n) * dS_h
+                if vorticity_type == 'absolute':
+                    Omega = as_vector((0, 0, self.parameters.Omega))
+                    L += inner(2*Omega, w) * dx
+            else:
+                vort = TrialFunction(space)
+                gamma = TestFunction(space)
+                n = FacetNormal(domain.mesh)
+                a = inner(vort, gamma) * dx
+                L = ( inner(domain.perp(grad(gamma)), u))*dx - jump(inner(domain.perp(n), u)*gamma)*ds_h
+                # TODO implement absolute version, unsure atm how to get corioilis in vertical slice smartly
```

Comment:
This isn't obvious to me, and it isn't obvious that we'll ever want the absolute vorticity in a vertical slice. I suggest that you put an if check here so that we fail with a `NotImplementedError` in this situation